examples/custom_rule.jl

```
@@ -0,0 +1,189 @@
+# and finally return a `Duplicated` for the output as promised. Let's see our rule in action! 
+# With the same setup as before:
+
+x  = [3.0, 1.0]
+dx = [1.0, 0.0]
+y  = [0.0, 0.0]
+dy = [0.0, 0.0]
+
+g(y, x) = f(y, x)^2 # function to differentiate
+
+@show autodiff(Forward, g, Duplicated(y, dy), Duplicated(x, dx)) # derivative of g w.r.t. x[1]
+@show dy; # derivative of y w.r.t. x[1] when g is run
+
+# We see that our custom forward rule has been triggered and gives the same answer as before.
+
+# ## Handling more activities 
+
+# Our custom rule applies for the specific set of activities that are annotated for `f` in the above `autodiff` call. 
+# However, Enzyme has a number of other annotations. Let us consider a particular case as an example, where the output
+# has a [`DuplicatedNoNeed`](@ref) annotation. This means we are only interested in its derivative, not its value.
+# To squeeze the last drop of performance, the below rule avoids computing the output of the original function and 
+# just computes its derivative.
+
+function forward(func::Const{typeof(f)}, ::Type{<:DuplicatedNoNeed}, y::Duplicated, x::Duplicated)
+    println("Using custom rule with DuplicatedNoNeed output.")
+    y.val .= x.val.^2 
+    y.dval .= 2 .* x.val .* x.dval
+    return sum(y.dval)
+end
+
+# Our rule is triggered, for example, when we call `autodiff` directly on `f`, as the output's derivative isn't needed:
+
+x  = [3.0, 1.0]
+dx = [1.0, 0.0]
+y  = [0.0, 0.0]
+dy = [0.0, 0.0]
+
+@show autodiff(Forward, f, Duplicated(y, dy), Duplicated(x, dx)) # derivative of f w.r.t. x[1]
+@show dy; # derivative of y w.r.t. x[1] when f is run
+
+# !!! note "Custom rule dispatch"
+#     When multiple custom rules for a function are defined, the correct rule is chosen chosen using 
+#     [Julia's multiple dispatch](https://docs.julialang.org/en/v1/manual/methods/#Methods).
+#     In particular, it is important to understand that the custom rule does not *determine* the
```

Is this correct? Was inspired by #695.

(And now that I think of it, if it's true, maybe there are some cases where we'd want custom activity annotations... e.g. if one writes `x, y = f(a, b)` where `f` is some function not understandable by Enzyme, and say the Jacobian of `f` is actually diagonal, it could be good to know that for activity analysis?)

---

Yes, it is correct. The one thing to add is "If a custom rule is specified for the correct function/argument types, but not the correct activity annotation, a runtime error will be thrown alerting the user to the missing activity rule rather than silently ignoring the rule."